perf(ngcc): reduce directory traversing

alan-agius4 opened the pull request:
This reduces the time that `findEntryPoints` takes from `9701.143ms` to `4177.278ms`, by reducing the file operations done.

Reference: #35717

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -80,17 +80,18 @@ export const SUPPORTED_FORMAT_PROPERTIES: EntryPointJsonProperty[] =
  * @param entryPointPath the absolute path to the potential entry-point.
- * @returns An entry-point if it is valid, `null` otherwise.
+ * @returns An entry-point if it is valid, `undefined` when it's not an entrypoint, `null`
+ * otherwise.
```
Could you be a bit more explicit by what "not an entry-point" and "otherwise" mean?
* `undefined`: there is no package.json at the path and there is no config to force an entry-point
* `null`: there is a package.json but it is not a valid Angular compiled entry-point

alan-agius4 replied:
Updated!

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts:
```diff
@@ -128,7 +128,7 @@ runInEachFileSystem(() => {
-      expect(entryPoint).toBe(null);
+      expect(entryPoint).toBe(undefined);
     });
 
```
Can we have a test for the case when `null` is returned?

alan-agius4 replied:
There is https://github.com/angular/angular/blob/5bc39f8c8d5238a9be9bd968cf18ea4b738bd6be/packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts#L58

petebacondarwin replied:
...except that you have changed that test to check for `undefined`!

alan-agius4 committed:
perf(ngcc): reduce directory traversing

This reduces the time that `findEntryPoints` takes from 9701.143ms to 4177.278ms, by reducing the file operations done.

Reference: #35717

alan-agius4 committed:
fixup! perf(ngcc): reduce directory traversing

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -113,26 +131,25 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
-          if (stat.isDirectory()) {
-            this.walkDirectory(packagePath, path, fn);
+          const containsEntryPoint = fn(absolutePath, stat.isDirectory());
+          if (containsEntryPoint) {
```
This check seems suspicious... isn't it possible to have an entry-point more than one directory below the package with no entry-point in between? E.g. `package/directory/entry-point`?

alan-agius4 replied:
I guess there is nothing to stop a user from creating such format, though I don't think I ever saw this.

alan-agius4 replied:
And in such case most likely it will get transformed by the sync version of the NGCC, when integrated with the CLI.

petebacondarwin replied:
OK let's go with this for now and revert if we get a complaint.

petebacondarwin reviewed (changes requested):
Thanks for this work @alan-agius4 - I can already see we will get benefits from this.

I get very nervous when we have code that starts using `null` and `undefined` to mean significant things. They are too magical and it is not easy to see immediately what they mean.

How about we modify the return value of these functions to be an object that has various flags to indicate the state of the result?

alan-agius4 committed:
fixup! perf(ngcc): reduce directory traversing

alan-agius4 commented:
Test for `null` -> https://github.com/angular/angular/blob/41b5b2714eb19a4dea334a92bccf1cb4ac66f214/packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts#L168

> I get very nervous when we have code that starts using null and undefined to mean significant things. They are too magical and it is not easy to see immediately what they mean.

Totally agree, though my worry would be that this will require some refactoring in the tests mainly. I'll look at this tomorrow.

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -40,66 +43,81 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
     if (entryPoints.length > 0) {
       // The `sourceDirectory` is an entry-point itself so no need to search its sub-directories.
+      // Also check for any nested node_modules in this package.
+      const nestedNodeModulesPath = this.fs.join(sourceDirectory, 'node_modules');
```
I was thinking of adding a check here, that we only try to traverse nested  `node_modules`, if one of the entrypoints is compiled by Angular.

It is very unlikely that a non Angular entry point has a dependency on an Angular library. 

Thoughts @petebacondarwin?

petebacondarwin replied:
I was thinking about this too. There is a very small chance, maybe, that someone creates a holding library that aggregates together a bunch of Angular libraries and re-exports them. But I think that this might well break ngcc anyway, since we would need to process the re-exported typings for it to work with ngtsc anyway.

So perhaps we can go with this idea.

alan-agius4 committed:
fixup! perf(ngcc): reduce directory traversing

alan-agius4 committed:
fixup! perf(ngcc): reduce directory traversing

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -11,9 +11,12 @@ import {Logger} from '../logging/logger';
 import {EntryPoint, getEntryPointInfo} from '../packages/entry_point';
 import {PathMappings} from '../utils';
+import {NGCC_DIRECTORY} from '../writing/new_entry_point_file_writer';
+
```
I know these extra blank lines are added by clang but you can remove them as they are uneeded, especially the one below.
